pennylane/transforms/optimization/commute_through_controls_targets.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transforms for pushing commuting gates through targets/control qubits."""
+
+from itertools import chain
+
+from pennylane import apply
+from pennylane.wires import Wires
+from pennylane.transforms import qfunc_transform
+
+from .optimization_utils import find_next_gate
+
+# Gates that commute among each particular basis
+commuting_gates = {
+    "X": ["PauliX", "RX", "SX"],
+    "Y": ["PauliY", "RY"],
+    "Z": ["PauliZ", "RZ", "PhaseShift", "S", "T"],
+}
+
+movable_gates = list(chain(*commuting_gates.values()))
+
+
+@qfunc_transform
+def commute_through_controls_targets(tape):
+    """Quantum function transform to move commuting gates past
+    control and target qubits of controlled operations.
+
+    Args:
+        qfunc (function): A quantum function.
+
+    **Example**
+
+    Consider the following quantum function.
+
+    .. code-block:: python
+
+        def qfunc(theta):
+            qml.CZ(wires=[0, 2])
+            qml.PauliX(wires=2)
+            qml.S(wires=0)
+
+            qml.CNOT(wires=[0, 1])
+
+            qml.PauliY(wires=1)
+            qml.CRY(theta, wires=[0, 1])
+            qml.PhaseShift(theta/2, wires=0)
+
+            qml.Toffoli(wires=[0, 1, 2])
+            qml.T(wires=0)
+            qml.RZ(theta/2, wires=1)
+
+            return qml.expval(qml.PauliZ(0))
+
+    >>> dev = qml.device('default.qubit', wires=3)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(0.5))
+     0: ──╭C──S──╭C─────╭C────────Rϕ(0.25)──╭C──T─────────┤ ⟨Z⟩
+     1: ──│──────╰X──Y──╰RY(0.5)────────────├C──RZ(0.25)──┤
+     2: ──╰Z──X─────────────────────────────╰X────────────┤
+
+    Diagonal gates on either side of control qubits do not affect the outcome
+    of controlled gates; thus we can push all the single-qubit gates on the
+    first qubit together on the right (and fuse them if desired). Similarly, X
+    gates commute with the target of ``CNOT`` and ``Toffoli`` (and ``PauliY``
+    with ``CRY``). We can use the transform to push single-qubit gates as
+    far as possible through the controlled operations:
```

Comment:
Question: could we have
```python
@commute_through_controls_target(direction="left"|"right")
```
in case you want to move the single-qubit operations in a particular direction? :D

Reply:
I love this!! After I wrote the transform I was still debating whether I should not rewrite it the other way (in fact, the precursors to it worked by pushing the controlled operations to the right and thus the single-qubit gates to the left, but didn't work nearly as well).

Also, this implicitly solves the long name problem :grin: